src/lib.rs

```diff
@@ -293,6 +300,18 @@ impl<'a> Iterator for Iter<'a> {
     }
 }
 
+/// Iterator over the string idtenfiers of the protocols (not addrs) in a multiaddr
+pub struct ProtoStackIter<'a> {
+    parts: Iter<'a>,
```

Comment:
Along similar lines, I did seriously consider implementing this in terms of a compact representation - perhaps u64 or u128, whose value is not valid on-the-wire... 
Upon construction eagerly iterate the protocols and for each one grab an ordinal, and pack _those_ into the representation (either bitwise or `*= COUNT; += ordinal`. Then only call `tag` when doing something like a Display.fmt. 
That would imply a `pub(crate)` version of `tag()` is actually matching on an ordinal, which the `pub` version could call after grabbing its enum's ordinal.

The idea would be to help out the other side - my metric `Family` could use this compact representation as a key rather than a Vec. 

But this might all be too much premature optimization, so I'm not going to go down that route unless someone else thinks it's worthwhile.

Reply:
I think this is a good idea. That said, I agree with your last paragraph. In other words, I only recommend going down this path in case we have a CPU or memory profile that shows that this is a bottleneck.